client/core/core.go

```diff
@@ -5080,14 +5080,40 @@ func processPreimageRequest(c *Core, dc *dexConnection, reqID uint64, oid order.
 		return fmt.Errorf("no active order found for preimage request for %s", oid)
 	}
 
+	checkCsum := func(csum dex.Bytes) error {
+		if csum != nil {
+			csumErr := errors.New("csum is already initialized")
+			resp, err := msgjson.NewResponse(reqID, nil,
+				msgjson.NewError(msgjson.InvalidRequestError, csumErr.Error()))
+			if err != nil {
+				return fmt.Errorf("preimage response encoding error: %w", err)
+			}
+			err = dc.Send(resp)
+			if err != nil {
+				return fmt.Errorf("preimage send error: %w", err)
+			}
+			return csumErr
+		}
+		return nil
+	}
+
 	// Record the commitment checksum so we can verify that the subsequent
 	// match_proof with this order has the same checksum. If it does not, the
 	// server may have used the knowledge of this preimage we are sending them
 	// now to alter the epoch shuffle.
+	//
+	// Allow to initialize csum only once per order to prevent possible
+	// malicious behavior.
 	tracker.mtx.Lock()
 	if isCancel {
+		if err := checkCsum(tracker.cancel.csum); err != nil {
+			return err
```

Comment:
The tracker.mtx is still locked if you return here.

Reply:
Good catch ;) Maybe there is already enough `csum`-relted logic to move it to a separate `func` ? 

Fixed by doing just that.